File: Medium/1143. Longest Common Subsequence/solution.cs
public class Solution {
    public int LongestCommonSubsequence(string text1, string text2) {
        var dp = new int[text1.Length, text2.Length];

        InitDP(text1, text2, dp);

        return Solve(0, 0, text1, text2, dp);
    }

    private int Solve(int i, int j, string text1, string text2, int[,] dp){
        if (i == text1.Length || j == text2.Length)
            return 0;

        if (dp[i, j] != -1)
            return dp[i, j];
        
        int result = 0;
        if (text1[i] == text2[j])
            result = 1 + Solve(i + 1, j + 1, text1, text2, dp);
        else
            result = Math.Max(Solve(i + 1, j, text1, text2, dp), Solve(i, j + 1, text1, text2, dp));

        dp[i, j] = result;
        return result;
    }

    private void InitDP(string text1, string text2, int[,] dp){
        for (int i = 0; i < text1.Length; i++){
            for (int j = 0; j < text2.Length; j++){
                dp[i, j] = -1;
            }
        }
    }
}

File: Medium/1143. Longest Common Subsequence/solution.cpp
class Solution {
public:

    int solve(int i, int j, string &text1, string &text2, vector<vector<int> > &dp){
        if (i == text1.size() || j == text2.size()) return 0;
        int &ret = dp[i][j];
        if (ret != -1)  return ret;

        if (text1[i] == text2[j]){
            ret = 1 + solve(i + 1, j + 1, text1, text2, dp);
        }
        else{
            ret = max(solve(i + 1, j, text1, text2, dp), solve(i, j + 1, text1, text2, dp));
        }
        return ret;
    }

    int longestCommonSubsequence(string text1, string text2) {
        int n = text1.size();
        int m = text2.size();

        vector<vector<int> > dp(n, vector<int> (m, -1));
        return solve(0, 0, text1, text2, dp);
    }
};